File: FinalAssignment/code/GreedyAlgorithm.py
import time

def read_hypergraph(filename):
    hypergraph = []
    with open(filename, 'r') as file:
        for line in file:
            edge = set(line.strip().split())
            hypergraph.append(edge)
    return hypergraph

def is_cover(hypergraph, vertex_cover):
    for edge in hypergraph:
        if not edge & vertex_cover:
            return False
    return True

def greedy_vertex_cover(hypergraph):
    cover = set()
    edges = set(map(frozenset, hypergraph))
    while edges:
        v = max(set().union(*edges), key=lambda x: sum(x in e for e in edges))
        cover.add(v)
        edges = {e for e in edges if v not in e}
    return cover

def save_vertex_cover(filename, vertex_cover):
    with open(filename, 'w') as file:
        file.write(" ".join(vertex_cover))

def main(filename, output_filename):
    hypergraph = read_hypergraph(filename)
    start_time = time.time()
    minimal_cover = greedy_vertex_cover(hypergraph)
    end_time = time.time()
    print(f"Time taken: {end_time - start_time:.4f} seconds")
    minimal_cover = sorted(list(minimal_cover), key=int)

    # Save minimal vertex cover to file
    save_vertex_cover(output_filename, minimal_cover)

    print("Size of  minimal vertex cover:", len(minimal_cover))

    return minimal_cover

# Example usage
hypergraph_filename = 'FinalAssignment/code/dataset_B/cryg10000.txt'
output_filename = 'FinalAssignment/code/dataset_B/minimal_cover.txt'
minimal_cover = main(hypergraph_filename, output_filename)
print("The minimal vertex cover is:", minimal_cover)

File: FinalAssignment/code/GreedyAlgorithm.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <sstream>
#include <algorithm>
#include <chrono>

using namespace std;
using namespace std::chrono;

vector<unordered_set<string>> read_hypergraph(const string &filename) {
	vector<unordered_set<string>> hypergraph;
	ifstream file(filename);
	string line;
	
	if (file.is_open()) {
		while (getline(file, line)) {
			istringstream iss(line);
			unordered_set<string> edge;
			string vertex;
			while (iss >> vertex) {
				edge.insert(vertex);
			}
			hypergraph.push_back(edge);
		}
		file.close();
	}
	return hypergraph;
}

bool is_cover(const vector<unordered_set<string>> &hypergraph, const unordered_set<string> &vertex_cover) {
	for (const auto &edge : hypergraph) {
		bool covered = false;
		for (const auto &v : vertex_cover) {
			if (edge.find(v) != edge.end()) {
				covered = true;
				break;
			}
		}
		if (!covered) return false;
	}
	return true;
}

unordered_set<string> greedy_vertex_cover(const vector<unordered_set<string>> &hypergraph) {
	unordered_set<string> cover;
	vector<unordered_set<string>> edges = hypergraph;
	
	while (!edges.empty()) {
		unordered_map<string, int> vertex_count;
		for (const auto &edge : edges) {
			for (const auto &v : edge) {
				vertex_count[v]++;
			}
		}
		
		auto max_vertex_it = max_element(vertex_count.begin(), vertex_count.end(),
			[](const pair<string, int> &a, const pair<string, int> &b) {
				return a.second < b.second;
			});
		
		string v = max_vertex_it->first;
		cover.insert(v);
		
		vector<unordered_set<string>> new_edges;
		for (const auto &edge : edges) {
			if (edge.find(v) == edge.end()) {
				new_edges.push_back(edge);
			}
		}
		edges = new_edges;
	}
	return cover;
}

void save_vertex_cover(const string &filename, const unordered_set<string> &vertex_cover) {
	ofstream file(filename);
	if (file.is_open()) {
		for (const auto &v : vertex_cover) {
			file << v << " ";
		}
		file.close();
	}
}

int main() {
	string hypergraph_filename = "D:/CodeSpace/C++/AlgorithmsHomeWork/FinalAssignment/code/dataset_B/circuit_3.txt";
	string output_filename = "D:/CodeSpace/C++/AlgorithmsHomeWork/FinalAssignment/code/dataset_B/minimal_cover.txt";
	
	vector<unordered_set<string>> hypergraph = read_hypergraph(hypergraph_filename);
	
	auto start_time = high_resolution_clock::now();
	unordered_set<string> minimal_cover = greedy_vertex_cover(hypergraph);
	auto end_time = high_resolution_clock::now();
	auto duration = duration_cast<microseconds>(end_time - start_time);
	cout << "Time taken: " << duration.count() / 1000000.0 << " seconds" << endl;
	
	vector<string> sorted_cover(minimal_cover.begin(), minimal_cover.end());
	sort(sorted_cover.begin(), sorted_cover.end(), [](const string &a, const string &b) {
		return stoi(a) < stoi(b);
	});
	
	// Save minimal vertex cover to file
	save_vertex_cover(output_filename, unordered_set<string>(sorted_cover.begin(), sorted_cover.end()));
	
	cout << "Size of minimal vertex cover: " << minimal_cover.size() << endl;
	cout << "The minimal vertex cover is: ";
	for (const auto &v : sorted_cover) {
		cout << v << " ";
	}
	cout << endl;
	
	return 0;
}

